Ponca/src/SpatialPartitioning/KdTree/kdTree.h

```diff
@@ -292,13 +274,136 @@ public :
+
+    // Internal ----------------------------------------------------------------
+protected:
+    inline KdTreeImplBase():
+        m_points(),
+        m_nodes(),
+        m_indices(),
+        m_min_cell_size(64),
+        m_leaf_count(0)
+    {
+    }
+
+    /// Generate a tree sampled from a custom contained type converted using a `Converter`
+    /// \tparam PointUserContainer Input point, transformed to PointContainer
+    /// \tparam IndexUserContainer Input sampling, transformed to IndexContainer
+    /// \tparam Converter
+    /// \param points Input points
+    /// \param sampling Indices of points used in the tree
+    /// \param c Cast/Convert input point type to DataType
+    template<typename PointUserContainer, typename IndexUserContainer, typename Converter>
+    inline void buildWithSampling(PointUserContainer&& points,
+                                  IndexUserContainer sampling,
+                                  Converter c);
+
+    /// Generate a tree sampled from a custom contained type converted using a \ref KdTreeImplBase::DefaultConverter
+    /// \tparam PointUserContainer Input points, transformed to PointContainer
+    /// \tparam IndexUserContainer Input sampling, transformed to IndexContainer
+    /// \param points Input points
+    /// \param sampling Samples used in the tree
+    template<typename PointUserContainer, typename IndexUserContainer>
+    inline void buildWithSampling(PointUserContainer&& points,
+                                  IndexUserContainer sampling)
+    {
+        buildWithSampling(std::forward<PointUserContainer>(points), std::move(sampling), DefaultConverter());
+    }
+
+private:
+    inline void build_rec(NodeIndexType node_id, IndexType start, IndexType end, int level);
+    inline IndexType partition(IndexType start, IndexType end, int dim, Scalar value);
+};
+
+/*!
+ * \brief Customizable base class for dense KdTree datastructure
+ *
+ * This version of the KdTree does not support subsampling. For an
+ * implementation that supports subsampling, see \ref KdTreeSparseBase.
+ *
+ * \see Ponca::KdTreeDense
+ * \see Ponca::KdTreeSparse
+ *
+ * \tparam Traits Traits type providing the types and constants used by the kd-tree. Must have the
+ * same interface as the default traits type.
+ *
+ * \see KdTreeDefaultTraits for the trait interface documentation.
+ */
+template <typename Traits>
+class KdTreeDenseBase : public KdTreeImplBase<Traits>
+{
+private:
+    using Base = KdTreeImplBase<Traits>;
+
+public:
+    /// Default constructor creating an empty tree
+    /// \see build
+    KdTreeDenseBase() = default;
+
+    /// Constructor generating a tree from a custom contained type converted using a \ref KdTreeImplBase::DefaultConverter
+    template<typename PointUserContainer>
+    inline explicit KdTreeDenseBase(PointUserContainer&& points)
+        : Base()
+    {
+        this->build(std::forward<PointUserContainer>(points));
+    }
+};
```

Comment:
Instead, couldn't we use a constructor in `KdTreeImplBase` and that could be called directly ? Base(std::forward<PointUserContainer>(points))`.
Then `KdTreeDenseBase` would only be an aliasing structure (that would be extended with upcoming PRs).

Reply:
Currently doing this would work, but when trying to implement inverse mapping I needed to call virtual functions from `build` which is not possible in the base constructor. I could change it for this PR but it would get changed again later

Reply:
Ok I see, so let's keep it like this.